pkg/querier/queryrange/log_result_cache.go

```diff
@@ -0,0 +1,280 @@
+package queryrange
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+	"github.com/weaveworks/common/httpgrpc"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logqlmodel/stats"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/tenant"
+	"github.com/grafana/loki/pkg/util/validation"
+)
+
+// LogResultCacheMetrics is the metrics wrapper used in log result cache.
+type LogResultCacheMetrics struct {
+	CacheHit  prometheus.Counter
+	CacheMiss prometheus.Counter
+}
+
+// NewLogResultCacheMetrics creates metrics to be used in log result cache.
+func NewLogResultCacheMetrics(registerer prometheus.Registerer) *LogResultCacheMetrics {
+	return &LogResultCacheMetrics{
+		CacheHit: promauto.With(registerer).NewCounter(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "query_frontend_log_result_cache_hit_total",
+		}),
+		CacheMiss: promauto.With(registerer).NewCounter(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "query_frontend_log_result_cache_miss_total",
+		}),
+	}
+}
+
+// NewLogResultCache creates a new log result cache middleware.
+// Currently it only caches empty filter queries, this is because those are usually easily and freely cacheable.
+// Log hits are difficult to handle because of the limit query parameter and the size of the response.
+// In the future it could be extended to cache non-empty query results.
+// see https://docs.google.com/document/d/1_mACOpxdWZ5K0cIedaja5gzMbv-m0lUVazqZd2O4mEU/edit
+func NewLogResultCache(logger log.Logger, limits Limits, cache cache.Cache, shouldCache queryrangebase.ShouldCacheFn, metrics *LogResultCacheMetrics) queryrangebase.Middleware {
+	if metrics == nil {
+		metrics = NewLogResultCacheMetrics(nil)
```

Comment:
I think it's probably because somewhere we make calls to `someMetric.Add/Inc/Set` etc. which would cause issues if the metric struct is nil. OTOH if the struct is instantiated but the `prometheus.Registerer` was nil then it's just like writing to `/dev/null`.

That being said, my opinion is that we should move to a pattern where our constructors either take in an already created metrics struct, or the `prometheus.Registerer` that will be passed to `NewTypeMetrics`.

Reply:
Callum is correct. I can pass prometheus.Registerer instead of nil if that fits better ? WDYT Callum ?

Reply:
I would lean towards the ctor `NewLogResultsCache` taking in an instnace of the metrics struct as a parameter, but passing a registerer in and using that is okay too.

Reply:
it is actually taking the instance of the metrics.